<img src="https://coursesandconferences.wellcomeconnectingscience.org/wp-content/themes/wcc_courses_and_conferences/dist/assets/svg/logo.svg" width="300" height="50"> 

# Genomics and Epidemiological Surveillance of Bacterial Pathogens - Latin America and the Caribbean

### 21–26 July 2024, San José, Costa Rica

[Wellcome Connecting Science Course Run Website Link](https://coursesandconferences.wellcomeconnectingscience.org/event/genomics-epidemiological-surveillance-of-bacterial-pathogens-latin-america-and-the-caribbean-20240721/) <br /> 
[Course Time Table 2024](https://github.com/WCSCourses/GenEpiLAC2024/blob/main/Manuals/Timetable.pdf) <br /> 
[Course Manual](https://github.com/WCSCourses/GenEpiLAC2024/blob/main/Manuals/Manual_main.md)

Learn how to apply genomics to epidemiological surveillance of bacterial pathogens and public health in Latin America

### Summary
In collaboration with the **[Universidad Nacional](https://www.una.ac.cr/)** and **[Universidad de Costa Rica](https://www.ucr.ac.cr/)**, we are pleased to announce our latest overseas course in Genomics and Epidemiological Surveillance of Bacterial Pathogens.

This course will provide training in the latest computational methods for analysis and interpretation of high throughput genome sequence data in a public health setting. Although the course is built around genomic surveillance the focus of this course is designed to reflect regional public health priorities on bacterial pathogens, with a strong emphasis on infections endemic to Latin America.

### Course background

Established in 2013, the course provides skills to combine epidemiological, metadata and genomics data analysis, hence contributing to generating the most reliable information available for public health decision making. In the last few years, the COVID-19 pandemic challenge accelerated the introduction of advances in genomics, sequencing technology and our ability to interpret high resolution sequence data for surveillance purposes. We are now seeing a shift from the use of molecular typing data for epidemiological surveillance and public health investigations of infectious disease to using Whole Genome Sequencing (WGS) for this purpose. In some settings WGS is now being used to define outbreaks and trace infection sources in humans, animals and the environment, contributing to reinforce a One Health approach. This approach also offers enormous promise for predicting antimicrobial resistance phenotypes.

Existing studies that have used genomics for mapping the spread of pathogens have highlighted two things: Firstly, genomic datasets are growing day by day and they offer the context against which to understand local patterns of disease. Secondly, these studies highlight the power gained from public health scientists and academic research scientists working together to better understand infectious diseases. This is the foundation of this course and expert instructors have been assembled from a wide range of settings in Latin America and Europe, including national and international public health organisations and front line high-profile genomic research organisations. We recognise the synergy between academic research and public health and this is your opportunity to strengthen this effort by attending the course.

### Target audience

The course is targeted at clinical or public health molecular biologists and microbiologists working in Latin America and the Caribbean.

### Programme

This course builds upon previous versions, incorporating genomics and problem solving exercises with modules that demonstrate how epidemiological information and phenotyping – in combination with sequence data – can be used:

1. for georeferencing (also known as ‘phylogeography’) where the location of disease causing strains are mapped with detailed genetic and phenotypic data (for example antibiotic resistance patterns) to look for regional patterns of disease, and
2. to understand the fine detail of new and emerging infections, the risk to public health and the management of infectious diseases.

The week-long programme will include:

- Practical bioinformatics skills needed to manipulate raw sequence data, including: mapping and de novo assemblies, SNP calling, annotation, antimicrobial resistance calling, and analysis of genomic information
- Principles of phylogeny and construction of accurate phylogenies
- Group exercises designed to teach everyday skills required to work in genomic surveillance
- Practical exercises for using GPS to map cases of infectious disease and phenotypes (such as AMR) across the world

### Learning outcomes

After attending the course, participants should be able to:

- Select and apply appropriate tools to analyse genome sequence data effectively
- Perform comparative genomics on multiple genome sequences
- Interpret phylogenetics to gain insights into pathogen evolution and epidemiology 
- Formulate projects using the global surveillance platform Microreact and Epicollect
- Explain, using examples, how bacterial pathogens and AMR spread among human populations.

**Course Instructors**

- [Caterina Guzmán-Verri](https://scholar.google.com/citations?user=S1GNAC8AAAAJ&hl=en), PIET, Veterinary School, Universidad Nacional, Costa Rica
- [Josefina Campos](https://www.researchgate.net/profile/Josefina-Campos-2), World Health Organisation, Switzerland
- [Matt Holden](https://risweb.st-andrews.ac.uk/portal/en/persons/matthew-holden(029e1abe-7829-4bc3-accf-9d9e392f274c).html), University of St Andrews, Scotland
- [Mathew Beale](https://www.sanger.ac.uk/person/beale-mathew/), Wellcome Sanger Institute, UK
- [Julio Caballero](https://www.pathogensurveillance.net/team/dr-julio-diaz-caballero/), Centre for Genomic Pathogen Surveillance, University of Oxford, UK
- [Matthew Dorman](https://www.universityofgalway.ie/our-research/people/mathematical-statistical-sciences/matthewdorman/), University of Galway, Ireland
- [Jordan Ashworth](), Public Health Scotland, UK
- [Francisco Duarte-Martinez](https://www.inciensa.sa.cr/investigacion/investigadores/Francisco%20Duarte.aspx#HERMES_TABS_1_0), INCIENSA, Costa Rica
- [Jose Arturo Molina Mora](https://ciet.ucr.ac.cr/en/researchers/jose-arturo-molina-mora-phd), Universidad de Costa Rica, Costa Rica
- [María Sol Haim](https://www.researchgate.net/profile/Maria-Sol-Haim), National Center of Genomics & Bioinformatics - ANLIS-Malbran, Argentina
- [Tomás Poklepovich](https://www.linkedin.com/in/tom%C3%A1s-javier-pokl%C3%A9povich-caride-6a622322/), National Center of Genomics & Bioinformatics - ANLIS-Malbran, Argentina
- [César Rodríguez](https://scholar.google.com/citations?user=Mq0-MIsAAAAJ&hl=en), Universidad de Costa Rica, Costa Rica
- [Marcela Suárez-Esquivel](https://www.researchgate.net/profile/Marcela-Suarez-Esquivel), Universidad Nacional, Costa Rica

**Wellcome Connecting Science Team**

- [Michelle Bishop](https://www.wellcomeconnectingscience.org/person/bishop-michelle/#), Associate Director of Learning and Training at Wellcome Connecting Science
- [Alice Matimba](https://uk.linkedin.com/in/alice-matimba-8805177), Head of Training and Global Capacity
- [Nicola Stevens](https://www.wellcomeconnectingscience.org/person/stevens-nicola/), Events Organiser
- [Isabela Malta](https://www.wellcomeconnectingscience.org/person/malta-isabela/), Assistant Overseas Courses Manager
- [Martin Aslett](https://coursesandconferences.wellcomeconnectingscience.org/about-us/the-team/), Informatics Manager
- [Vaishnavi Vikas Gangadhar](https://www.wellcomeconnectingscience.org/person/gangadhar-vaishnavi/), Informatics Technical Officer

******

### Citing and Re-using Course Material

The course data are free to reuse and adapt with appropriate attribution. All course data in these repositories are licensed under the <a rel="license" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">Attribution-NonCommercial-ShareAlike 4.0 International (CC BY-NC-SA 4.0)</a>. <a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br /> 

Each course landing page is assigned a DOI via Zenodo, providing a stable and citable reference. These DOIs can be found on the respective course landing pages and can be included in CVs or research publications, offering a professional record of the course contributions.

### Interested in attending a course?

Take a look at what courses are coming up at [Wellcome Connecting Science Courses & Conference Website](https://coursesandconferences.wellcomeconnectingscience.org/our-events/).

---

For more information or queries, feel free to contact us via the [Wellcome Connecting Science website](https://coursesandconferences.wellcomeconnectingscience.org).<br /> 
Find us on socials [Wellcome Connecting Science Linktr](https://linktr.ee/eventswcs)

---

[Wellcome Connecting Science GitHub Home Page](https://github.com/WCSCourses) <br /> 


******
## License
<a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">Attribution-NonCommercial-ShareAlike 4.0 International (CC BY-NC-SA 4.0)</a>.
