<img src="https://coursesandconferences.wellcomeconnectingscience.org/wp-content/themes/wcc_courses_and_conferences/dist/assets/svg/logo.svg" width="200" height="200">


[<<< Go back to Manual Contents Page](https://github.com/WCSCourses/GenEpiLAC2023/blob/main/Manuals/README.md)

<br>

# Genome Assembly and Analysis - Paraguay 2023 <!-- omit in toc -->

### Module Leads: Mat Beale and Jordan Ashworth<!-- omit in toc -->

<br>

# Table of contents <!-- omit in toc -->
- [Introduction](#introduction)
  - [Short-read sequencing assembly](#short-read-sequencing-assembly)
  - [Long-read sequencing assembly](#long-read-sequencing-assembly)
  - [Hybrid Assembly](#hybrid-assembly)
  - [What assemblies are best?](#what-assemblies-are-best)

<br>


# [Introduction](#introduction)

## [Short-read sequencing assembly](#short-read-sequencing-assembly)

The data from the Illumina machine comes as relatively short stretches (35-250 bp) of DNA â€“ around 6 billion of them for a typical bacterial genome. These individual sequences are called sequencing reads. There are a range of assembly programs that have been specifically designed to assemble genomes from next-generation sequence (NGS) data. Genome assembly using sequence reads of around 100bp is complicated due to the high frequency of repeats longer than the sequence read length in genomes, such as insertion sequence (IS) elements and rRNA operons, and the massive amount of data the programs have to handle. In addition to finding overlaps in the sequence, the assembly programs can also use information from the predicted insert size when paired reads are used, to link and position reads in an assembly.

When a genome is pieced together without any reference sequence to compare it to or scaffold against, it is termed a _de novo_ assembly. Due to the challenges of assembly mentioned earlier, de novo assembly using short reads will not produce complete genomes, but rather fragmented assemblies of multiple contiguous sequences (contigs), the order of which is arbitrary and does not necessarily reflect their actual order in the genome.

## [Long-read sequencing assembly](#long-read-sequencing-assembly)

With the advent of real-time sequencing technologies like Oxford Nanopore and Pacific Biosciences, it is now possible to generate sequence reads that far exceed the size of those produced by Illumina platforms. These long-read technologies can produce reads tens of thousands of base pairs in length, which contrasts with the tens or hundreds of base pairs generated by Illumina platforms.

The longer reads from these technologies can span through repeat regions that limit the assembly of short-read sequence data. However, the higher error rates associated with some long-read sequencing technologies and the variable size distribution of reads they generate mean that generating high-quality complete assemblies from long-reads alone can be challenging.

## [Hybrid Assembly](#hybrid-Assembly)

A popular approach to generate more complete genomes is to combine the accuracy of short-read sequencing data with the long-range information provided by long-read sequencing data, known as hybrid assembly. This approach requires sequencing an isolate using both short-read and long-read technologies. Assembler programs designed for hybrid assembly can utilize both types of data to find overlaps, stitch sequences together, and accurately predict the sequence in the assembly.

## [What assemblies are best?](#what-assemblies-are-best?)

Hybrid assemblies that achieve greater completeness offer distinct advantages for downstream analysis, such as understanding mobile genetic elements (MGEs), genome rearrangements, and evolutionary studies. However, these assemblies require dual sequencing technologies and entail additional costs, making them impractical for all isolates.

For tasks focused on genotyping and identifying specific genetic markers like antibiotic resistance genes, a draft genome from short-read data may suffice, avoiding the complexities and costs of hybrid approaches. The choice of assembly strategy should align closely with research objectives, balancing the need for comprehensive genomic information with practical considerations.

<br>
